### PR TITLE
spectral-ip-paper-2018

### PR DESCRIPTION
**Pull request by sgkang**
- consideration of repetitive waveform
- need to add examples

### CODE REVIEW

File: SimPEG/Maps.py [hunk 1]
```diff
@@ -1494,7 +1494,7 @@ def deriv(self, m, v=None):
 #                                                                             #
 #                             Parametric Maps                                 #
 #                                                                             #
-###############################################################################
+######################s#########################################################
```
Comment (lheagy) [inline SimPEG/Maps.py]:
```suggestion
###############################################################################
```

File: SimPEG/EM/Static/SIP/ProblemSIP.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 from SimPEG.EM.Static.IP import Problem3D_N as BaseProblem3D_N
 from .SurveySIP import Survey, Data
 import gc
+# from profilehooks import profile
```
Comment (lheagy) [inline SimPEG/EM/Static/SIP/ProblemSIP.py]:
```suggestion
```

File: SimPEG/EM/Static/SIP/ProblemSIP.py [hunk 3]
```diff
@@ -215,6 +430,7 @@ def forward(self, m, f=None):
 
             return self.sign*np.hstack(Jv)
 
+    # @profile
```
Comment (lheagy) [inline SimPEG/EM/Static/SIP/ProblemSIP.py]:
```suggestion
```

File: SimPEG/EM/Static/SIP/ProblemSIP.py [hunk 4]
```diff
@@ -259,18 +475,17 @@ def Jvec(self, m, v, f=None):
...
+                        # timeindex = rx.getTimeP(self.survey.times)
+                        # if timeindex[tind]:
+                        Jv_temp = (
+                            rx.evalDeriv(src, self.mesh, f, du_dm_v)
+                            )
+                        Jv.append(Jv_temp)
 
             return self.sign*np.hstack(Jv)
 
+    # @profile
```
Comment (lheagy) [inline SimPEG/EM/Static/SIP/ProblemSIP.py]:
```suggestion
```

**Comment by sgkang** [comment]
Hi @lheagy, this repo should be good to go. Will be great if we can merge to simulation refactor.

File: SimPEG/EM/Static/DC/ProblemDC_2D.py [hunk 5]
```diff
@@ -560,7 +562,7 @@ def getA(self, ky):
         # This seems not required for 2.5D problem
         # Handling Null space of A
         # A[0, 0] = A[0, 0] + 1.
-        # print(A.shape, 'N')
+        # print (A.shape, 'N')
```
Comment (lheagy) [inline SimPEG/EM/Static/DC/ProblemDC_2D.py]:
```suggestion
```

File: SimPEG/EM/Static/IP/ProblemIP.py [hunk 6]
```diff
@@ -13,7 +13,8 @@
 from .SurveyIP import Survey
 from SimPEG import Props
 import sys
-
+import scipy.sparse as sp
+# from profilehooks import profile
```
Comment (lheagy) [inline SimPEG/EM/Static/IP/ProblemIP.py]:
```suggestion
```

File: SimPEG/EM/Static/IP/ProblemIP.py [hunk 7]
```diff
@@ -114,6 +129,10 @@ def Jvec(self, m, v, f=None):
             # Resistivity (d u / d log rho) - HJ form
             return self.sign*np.hstack(Jv)
 
+    def forward(self, m, f=None):
+        return self.Jvec(m, m, f=f)
+
+    # @profile
```
Comment (lheagy) [inline SimPEG/EM/Static/IP/ProblemIP.py]:
```suggestion
```

**Comment by lheagy** [comment]
:wave: @seogi, would you mind updating the main pr thread above to give a short description of the changes? I will use this for the release notes. Thanks!

**Comment by sgkang** [comment]
Spectral Induced Polarization Inversion
--------------------------------------------

- Invert multiple time channels of IP data to recover 3D distribution of polarization 
parameters: `eta`, `tau`, and `c`

- Can handle various data types: `volt` (V) or `apparent_chargeability` (V/V)

- Profile `SIP` modules to make sure it can handle large-scale time-domain IP data (`storeJ=False`)

image

DC and IP codes
------------------

- Can handle various data types: `volt` (V) or `apparent_resistivity` (ohm-m), or `apparent_chargeability` (V/V)

Examples using `DC`, `IP`, `SIP` codes are available through:

https://github.com/simpeg-research/kang-2018-spectral-inducedpolarization/tree/master/notebooks

**Pull request merged**